### ISSUE TITLE
Reload dialog, choose "No", file reloads

### ISSUE THREAD

**Issue by dcorsello** [notepad-plus-plus/notepad-plus-plus#10816]
### Description of the Issue

When notepad++ detects that an open file has been updated by another program, the reload dialog appears, asking if I want to reload the file and lose changes.  When I choose "No", the file reloads anyway, and I lose all changes.  This is a HUGE issue that has resulted in much lost work!  I'm using v8.1.9.2.

### Steps to Reproduce the Issue

1. Open the NppFTP window and double-click on a file to open it from the server for editing.
2. Edit the file and do not save.
3. Double-click on the same file in the NppFTP window.
4. When the Reload dialog appears, click on "No".

### Expected Behavior

The file should not reload from the server, and the current editing history should be preserved.

### Actual Behavior

The file reloads from the server, all unsaved changes are lost, and the current editing history is lost.

### Debug Information

Notepad++ v8.1.9.2   (32-bit)
Build time : Nov 21 2021 - 04:27:12
Path : C:\Program Files (x86)\Notepad++\notepad++.exe
Command Line : 
Admin mode : OFF
Local Conf mode : OFF
Cloud Config : OFF
OS Name : Windows 10 Enterprise (64-bit) 
OS Version : 2009
OS Build : 19043.1348
Current ANSI codepage : 1252
Plugins : ComparePlugin.dll DSpellCheck.dll mimeTools.dll NppConverter.dll NppExport.dll NppFTP.dll 

Sometimes it is necessary for me to open a number of files from the server, make changes to them, and save all of them simultaneously.  In past versions of Notepad++, if I was working this way and accidentally tried to reopen a file from the server, clicking "No" in the Reload dialog was a way to continue working.  Now, it's an absolute disaster!  This is basic, essential functionality that is now broken.  Please fix ASAP.

**Comment by dcorsello** [comment]
I did a clean install of v8.1.9.2 and the problem continues.

**Comment by dcorsello** [comment]
I also tried v8.1.5, v8.1.4 and 7.9.5, all with no success.  It would seem that the problem is outside of Notepad++, but the following two circumstances seem to me to make this unlikely: 1) No software was installed, updated or uninstalled on my computer besides Windows Updates.  2) The most recent Windows Update was a Windows Defender definitions update yesterday, and it seems unlikely to me that this would result in a functionality change in Npp. 3) The most recent operating system updates were installed 12 days ago. and the problem I'm reporting started yesterday.

Although I chose to delete all settings on each uninstall, I suppose it's possible that the uninstalls of v8.1.9.2, v8.1.5 and v8.1.4 were incomplete, so that downgrading Npp had no effect.

**Comment by dcorsello** [comment]
It's not computer-specific.  I just installed Npp on another Windows 10 machine, and got the same unexpected behavior.  This is a showstopper for me.

**Comment by dcorsello** [comment]
Notepad++ has been a wonderful tool for years.  I see it's the fourth most popular IDE among professional developers according to the 2021 Stack Overflow Annual Developer Survey.  But this couldn't have happened at a worse time for me.

**Comment by xomx** [comment]
I can reproduce the described behavior (8.1.9.2 x64, NppFTP 0.29.8, Win8.1Ent, Synology NAS, SFTP).

But why are you trying to reopen the same file from the network storage?

And this could be also a NppFTP plugin issue (I did not check that yet).

**Comment by xomx** [comment]
I debugged it and it seems to me as a NppFTP issue indeed, as the N++ respect correctly the "No" answer and it handles the opened file buffer accordingly, but then the NppFTP plugin sends NPPM_RELOADFILE message to N++ anyway and so the N++ has to obey...

So this issue should go here:
https://github.com/ashkulz/NppFTP/issues

**Comment by dcorsello** [comment]
Thank you for your replies.  I opened an issue with NppFTP.  I would never intentionally reopen a file from the server.  But sometimes I do it accidentally, usually when I have a large number of files open.

**Comment by dcorsello** [comment]
Perhaps requiring N++ to always obey plugins (I think that's what you're implying) should be reconsidered.  In a case like this, in which a plugin compromises basic functionality, the effect can be disastrous.

**Comment by xomx** [comment]
Mission impossible ;-)

The installed N++ plugins are in fact an integral part of the running N++ process. The code running in a N++ plugin has the same rights as the code in the main N++ application. Theoretically any obstacle in between the N++ and a potentially evil plugin can be overcome.

And imagine that you are the maintainer of the N++ source code - should you then check source codes of every existing N++ plugin for a possible bug? And what about the closed source plugins?

**Comment by dcorsello** [comment]
My suspicion is that this is a N++ problem because the problem started right after a N++ update.  Also, I've installed versions of NppFTP as old as v0.27.3 from December, 2017 without success.  Also, the section of code that you recommended editing has remained unchanged for years.

Nevertheless, I used Visual Studio to modify the .dll per your suggestions.  But N++ complains that the new NppFTP.dll isn’t Unicode.  Any suggestions?  (I'm a web developer, not a c++ programmer.)

**Comment by xomx** [comment]
https://github.com/ashkulz/NppFTP/issues/327

**Comment by rdipardo** [comment]
@dcorsello,
> N++ complains that the new NppFTP.dll isn’t Unicode. Any suggestions?

I know this was resolved by https://github.com/ashkulz/NppFTP/issues/327#issuecomment-982643156, but here's the long answer to your original question.

If you look at the  [plugin template project](https://github.com/npp-plugins/plugindemo/blob/master/src/NppPluginDemo.cpp), there's a standard interface that every N++ plugin must provide, i.e.:

```cpp
  /* 1. */  BOOL isUnicode();
  /* 2. */  void setInfo(NppData notpadPlusData);
  /* 3. */  const TCHAR * getName();
  /* 4. */  FuncItem * getFuncsArray(int *nbF);
  /* 5. */  void beNotified(SCNotification *notifyCode);
  /* 6. */  LRESULT messageProc(UINT /*Message*/, WPARAM /*wParam*/, LPARAM /*lParam*/);
```

Most functions can be implemented with boilerplate [like this](https://github.com/ashkulz/NppFTP/blob/5e60962d1b6921b8707c71fd00a1ada3ff560f2b/src/PluginInterface.cpp#L145):

```cpp
extern "C" __declspec(dllexport) BOOL isUnicode()
{
    return TRUE;
}
```

Sounds like your patched version of `PluginInterface.cpp` was not exporting the `isUnicode()` function, probably because the `UNICODE` definition was not set in your project configuration. The "Properties" menu in VS should look like this under "Configuration Properties" > C/C++ > Preprocessor > "Preprocessor Definitions":

![Screenshot 2021-11-30 213905](https://user-images.githubusercontent.com/59004801/144162037-e91f715a-7eb4-4808-83f8-14984f270617.png)

**Comment by dcorsello** [comment]
@rdipardo 

Thanks a lot for that info.

**Comment by dcorsello** [comment]
@xomx 

Well, it seemed that your patch had fixed the problem, but I'm sorry to report that it just happened with your v0.29.8.1.  Fortunately, I didn't lose any work, but a file that had been saved got marked as unsaved after accidentally attempting to reload via NppFTP.

**Comment by xomx** [comment]
@dcorsello 
> but a file that had been saved got marked as unsaved after accidentally attempting to reload via NppFTP
I can reproduce that with my patched NppFTP.dll.
Maybe this is similar to #10796 , investigating now...

**Comment by xomx** [comment]
@dcorsello 
So it is not a bug as in the issue linked above.

You triggered the reloading by clicking on the same file in the NppFTP plugin subwindow.
Then you decided not to reload the file, because of it was an accident.
N++ set its relevant filebuffer as unsynced with the source (ie as dirty = unsaved) because of the filebuffer and its source file on disk can be different now (this is the current intended behavior of N++).

Explanation - there is no magic telling the N++ that your doubleclick on file in the NppFTP subwindow is intentional or not...

**Comment by donho** [comment]
So it's rather an issue of NppFTP but not one of Notepad++, isn't it?

**Comment by dcorsello** [comment]
@xomx 

Marking a saved file as dirty in this case is new, unexpected and confusing behavior in N++.  I don't think N++ needs to know my intention.  In my opinion, it just needs to honor my "no" as it always has.  Although the filebuffer and file on disk might be different, N++ shouldn't arbitrarily decide that they are different.  My "no" makes whether or not they are different a moot point.  In other words, N++ needs to preserve the file's current clean or dirty status and consider it changed only if I answer "yes".  Doesn't N++ know this status at the moment I try to reopen it with NppFTP?

If the current behavior is intended, then please consider this an urgent feature request to revert to the old behavior.  

From a user experience perspective, the end result of this new behavior is that if I'm slamming with 50 files open, and a file unexpectedly gets marked as changed, and I don't notice right away, it's going to make me nervous.  I shouldn't be forced to mentally keep track of all the files I've changed or haven't changed in a work session.  The software must accurately reflect the state of each file.

Thankfully,  your patched plugin preserves pending changes and edit history with a "no" response.

**Comment by xomx** [comment]
@dcorsello 

I understand, what you are saying and I agree that this behavior is sub-optimal. But this behavior is not new in N++, it is there a long time because of it is a part of the file-change detection mechanism. You can verify it by:

1. Open a file in N++.
2. Edit the same file in a different editor and save it there.
3. Switch back to N++ and there will be the changed-file detection message with the option to reload / not reload the file from its source on disk.
4a. When you decide to reload - file is reloaded  (then the file status in N++ is set as not dirty = unchanged, because of now the filebuffer in N++ and its source file on disk are the same for sure)
4b. When you decide not to reload - filebuffer in N++ is left as is, but it is automatically marked as dirty because of now the filebuffer and its source file on disk does not match.

And you unfortunately mimic that stuff in N++.

**Comment by xomx** [comment]
@donho 
There are 2 different things is this issue:

1) 
The original reported one (accidental reload of the same file from within the NppFTP subwindow even if the user says 'No' in the reload N++ messagebox) is most probably the NppFTP plugin issue, but I still do not have time to complete the investigation (more in the https://github.com/ashkulz/NppFTP/issues/327). 

2) 
After the above problem has been fixed by my patched version of NppFTP, the user reported less severe one, but I found that it is in fact sub-optimal, but standard N++ behavior.

And there is also a 3rd similar problem elsewhere (#10796 ), which seems to me as a real bug in N++ and I am going to do a PR for it.

**Comment by dcorsello** [comment]
@xomx 

Okay, so now I understand the mechanism.  As soon as I double-click on a file in NppFTP, it is downloaded to disk.  My "no" response in the Reload dialog doesn't prevent the download (I might have subconsciously expected that it would prevent it).  So, the reload dialog is accurately triggered.  But I don't recall seeing the status change to dirty in N++ under these circumstances before.  So, either something has changed, or my recollection is incorrect, clouded by losing edit history in unpatched NppFTP.  At least with your patched plugin I seem to be able to count on the preservation of edit history.  If a file gets marked as dirty after an accidental NppFTP download, I can do an edit undo to see if I actually made changes.

I think probably the only solution would be for NppFTP to give the user an opportunity to cancel the download of a file that's already open in N++.

**Comment by xomx** [comment]
@dcorsello 

> As soon as I double-click on a file in NppFTP, it is downloaded to disk. My "no" response in the Reload dialog doesn't prevent the download

Yes, that is the real mechanism behind. The NppFTP creates its own cache on the local disk from the files doubleclicked by you in the NppFTP subwindow. This mechanism is used there to reduce network traffic and probably also to overcome some of the FTP protocol limitations.



> If a file gets marked as dirty after an accidental NppFTP download, I can do an edit undo to see if I actually made changes

For surety I checked the real file state after it is marked so as dirty and it is really only a mark - the real content of the file is in fact still the same as before (I saved the N++ filebuffer marked in this way as dirty to a different file for a later comparison....).


> I think probably the only solution would be for NppFTP to give the user an opportunity to cancel the download of a file that's already open in N++. This seems reasonable, but I'm not sure how responsive the NppFTP team is.

Yes, that needs additional coding in the NppFTP. We will see. Never say never :-)


> Thanks for your thorough pursuit of this issue.

You're welcome. And this helped me as well, I am using the NppFTP too.

**Comment by donho** [comment]
@xomx Is this issue fixed in v8.1.9.3 ? (Fix dirty status persisting issue after reloading file.)

**Comment by xomx** [comment]
@donho @dcorsello 

The additional problem described here was not influenced by #10796 .

And the original problem reported here is a NppFTP issue, so this N++ issue should be closed.

Probably a PR for this issue will be in the NppFTP project, the project maintainer also thinks that this is a problem of the NppFTP: https://github.com/ashkulz/NppFTP/issues/327#issuecomment-987212893
